add doc for ci/jenkins (#7)
antrea-io/nephe: ci/jenkins/README.md

# Nephe CI: Jenkins

This directory includes all the scripts required to run CI for nephe.

> We have static VM as jenkins node, when there is a pull request or daily build trigger the job. The code will be cloned on

Reviewer: Can we truncate the line width to 80.

Author: done

> We have static VM as jenkins node, when there is a pull request or daily build trigger the job. The code will be cloned on

Reviewer: ```suggestion
We have a static VM configured to run as jenkins node.  When a jenkins job is triggered, the nephe code will be cloned on the static VM.
```

Author: done

> static VM, and it will call terraform to deploy dynamic VM and use ansible for system configuration such as installing docker,

Reviewer: ```suggestion
static VM. The jenkins node will call terraform to a deploy dynamic VM and use ansible for installing any missing packages like docker.
```

Author: done

and then run tests on the dynamic VM, destroy the VM once the test succeeds or fails.

> ## Terraform

Author: done

> We create terrform tfvars file in nephe-ci.sh with jenkins credetials. `deploy.sh` is the script to create vm, and store the

Reviewer: Please add hyperlinks when you are referring to scripts.
For example: [deploy.sh](deploy.sh)

Author: done

terraform state in a specified directory, which can be used to get VM information later. `destroy.sh` helps to detroy the VM
and delete terraform workspace as well.

For more details about terraform, please refer [here](https://www.terraform.io/docs)

## Ansible
Ansible will help us for system configuration after the VM installation finishes, all ansible tasks are in playbook/roles.
For example, we add task to install docker in [playbook/roles/docker/tasks/main.yml](playbook/roles/docker/tasks/main.yml).

Ansile also help to copy ssh keys to dynamic VM, by which we can run tests and collect log later.
The ssh-keys are generated on static VM if nephe-ci checks there is no ssh-keys there.
## jenkins job
We use jenkins job-builder to manage jenkins jobs, the codes are in the directory jobs. There are four files:

> * **job-templates.yaml**: If you need several jobs defined that are nearly identical, except perhaps in their names,

Reviewer: This may not be required.   Can you please take a look at https://github.com/antrea-io/antrea/blob/main/ci/jenkins/README.md as reference.

So it would look something like this.

# Introduction

# Jenkins Deployment

  ## Teraform
  
  ## Ansible
  
  ## Jenkins Job
  
  ### Requirements
  
  ### Jenkins job updater
 
# List of jenkins job

Author: done

# Nephe CI: Jenkins

## Table of Contents

<!-- toc -->
- [Introduction](#introduction)
- [Jenkins Deployment](#jenkins-deployment)
  - [Teraform](#teraform)
  - [Ansible](#ansible)
  - [Jenkins Job](#jenkins-job)
    - [Requirements](#requirements)
    - [Apply the jobs](#apply-the-jobs)
    - [Jenkins job updater](#jenkins-job-updater)
- [List of jenkins job](#list-of-jenkins-jobs)
<!-- /toc -->

## Introduction

This directory includes all the scripts required to run CI for nephe.

We have a cluster of static VMs configured to run as jenkins node.  When a jenkins
job is triggered, the nephe code will be cloned on the static VM. The jenkins node

> will call terraform to a deploy dynamic VM and use ansible for installing any missing

Reviewer: ```suggestion
will call terraform to deploy a dynamic VM and use ansible for installing any missing
```

Author: done

> packages like docker. Then tests will be run on that dynamic VM, Upon completion

Reviewer: ```suggestion
packages like docker. The tests will be run on the dynamic VM and upon completion, the dynamic VM will be destroyed.
```

of the test (success or fail), the dynamic VM will be destroyed.

## Jenkins Deployment

### Teraform

> [nephi-ci.sh](./nephe-ci.sh) is responsible for creating terraform tfvars files.

Reviewer: ```suggestion
- [nephi-ci.sh](./nephe-ci.sh) is responsible for creating terraform tfvars files.
```

Author: done

> [deploy.sh](./deploy.sh)is the script to create vm, and store the terraform state

Reviewer: ```suggestion
- [deploy.sh](./deploy.sh) is the script to create dynamic VM, and store the terraform state
```

Author: done

> in a specified directory, which can be used to get VM information later.

Reviewer: ```suggestion
in a specified directory, which can be used to retrieve VM information.
```

Author: done

> [destroy.sh](./destroy.sh) helps to detroy the VM and delete terraform workspace as well.

Reviewer: ```suggestion
- [destroy.sh](./destroy.sh) helps to destroy the dynamic VM and delete terraform workspace as well.
```

Author: done

> [destroy.sh](./destroy.sh) helps to detroy the VM and delete terraform workspace as well.

Reviewer: I don't understand usage of terraform workspace. 
Can you mention what it means in the terraform section and how is the CI using it?

> For more details about terraform, please refer [here](https://www.terraform.io/docs).

Reviewer: ```suggestion
Pease refer to [terraform](https://www.terraform.io/docs) documentation for more information.
```

Author: done

### Ansible

Ansible is used to configure VM once terraform deploys the dynamic VM. All anisble
tasks are in playbook/roles. For example, we add task to install docker in
[playbook/roles/docker/tasks/main.yml](playbook/roles/docker/tasks/main.yml).

Ansible is used to copy ssh keys to the dynamic VM, which is used to run the tests
and collect logs from the dynamic VM later.

> For more details about ansible, please refer [here](https://docs.ansible.com/).

Reviewer: Please 
```suggestion
Please refer to [ansible](https://docs.ansible.com/) documentation for more information.
```

Author: done

<!-- toc -->
- [Introduction](#introduction)
- [Jenkins Deployment](#jenkins-deployment)
  - [Teraform](#teraform)
  - [Ansible](#ansible)
  - [Jenkins Job](#jenkins-job)
    - [Requirements](#requirements)
    - [Apply the jobs](#apply-the-jobs)
    - [Jenkins job updater](#jenkins-job-updater)
- [List of jenkins job](#list-of-jenkins-jobs)
<!-- /toc -->

## Introduction

This directory includes all the scripts required to run CI for nephe.

We have a cluster of static VMs configured to run as jenkins node.  When a jenkins
job is triggered, the nephe code will be cloned on the static VM. The jenkins node
will call terraform to deploy a dynamic VM and use ansible for installing any missing
packages like docker. Then tests will be run on that dynamic VM, Upon completion
of the test (success or fail), the dynamic VM will be destroyed.

## Jenkins Deployment

### Teraform

The persistent data stored in the backend belongs to a workspace. Named workspaces allow
conveniently switching between multiple instances of a single configuration within its single
backend. Multiple workspaces are used here to create multiple dynamic VMs in parallel, each
dynamic VM is in a workspace.

> -[nephi-ci.sh](./nephe-ci.sh) is responsible for creating terraform tfvars files.

Reviewer: - [nephi-ci.sh](./nephe-ci.sh) is responsible for creating terraform tfvars files.
- [deploy.sh](./deploy.sh) is the script to create vm, and store the terraform state
   in a specified directory, which can be used to retrieve VM information.
- [destroy.sh](./destroy.sh) helps to destroy the dynamic VM and delete terraform workspace as well.

Author: done

-[deploy.sh](./deploy.sh) is the script to create vm, and store the terraform state
in a specified directory, which can be used to retrieve VM information.
-[destroy.sh](./destroy.sh) helps to destroy the dynamic VM and delete terraform workspace as well.

Please refer to [terraform](https://www.terraform.io/docs) documentation for more information.

### Ansible

Ansible is used to configure VM once terraform deploys the dynamic VM. All anisble
tasks are in playbook/roles. For example, we add task to install docker in
[playbook/roles/docker/tasks/main.yml](playbook/roles/docker/tasks/main.yml).

Ansible is used to copy ssh keys to the dynamic VM, which is used to run the tests
and collect logs from the dynamic VM later.

Please refer to [ansible](https://docs.ansible.com/) documentation for more information.

### Jenkins Job

> We use jenkins job-builder to manage jenkins jobs, the codes are in the [directory jobs](./jobs).

Reviewer: ```suggestion
We use jenkins job-builder to manage jenkins jobs and the jenkins yamls are placed in the directory [jobs](./jobs).
```

Author: done

#### Requirements

Jenkins jobs can be updated through yaml files in [jobs](./jobs).

> Yaml files under current directory can be generated via jenkins-job-builder.

Reviewer: ```suggestion
Yaml files under current directory can be generated via `jenkins-job-builder`.
```

Author: done

If you want to verify the changes in yaml files, please install
[jenkins-job-builder](https://jenkins-job-builder.readthedocs.io/en/latest/index.html)

> #### test the jobs

Reviewer: Validate Jobs?

Author: done

[deploy.sh](./deploy.sh)is the script to create vm, and store the terraform state
in a specified directory, which can be used to get VM information later.
[destroy.sh](./destroy.sh) helps to detroy the VM and delete terraform workspace as well.

For more details about terraform, please refer [here](https://www.terraform.io/docs).

### Ansible

Ansible is used to configure VM once terraform deploys the dynamic VM. All anisble
tasks are in playbook/roles. For example, we add task to install docker in
[playbook/roles/docker/tasks/main.yml](playbook/roles/docker/tasks/main.yml).

Ansible is used to copy ssh keys to the dynamic VM, which is used to run the tests
and collect logs from the dynamic VM later.

For more details about ansible, please refer [here](https://docs.ansible.com/).

### Jenkins Job

We use jenkins job-builder to manage jenkins jobs, the codes are in the [directory jobs](./jobs).

#### Requirements

Yaml files under current directory can be generated via jenkins-job-builder.
If you want to try out the tests on your local jenkins setup, please check
the following requirements:

* Jenkins setup
  * Plugins: ghprb, throttle-concurrents

>   * Jenkins IP/User/Token: they are store in jenkins, and exported to test scripts.

Reviewer 2: What exactly needs to be set to point to private jenkins.. Can you add sample, or exact env variable here

* Install
  [jenkins-job-builder](https://jenkins-job-builder.readthedocs.io/en/latest/index.html)
* Define your `ANTREA_GIT_CREDENTIAL` which is the credential for your private
  repo.
* Define your `ghpr_auth`, `nephe_admin_list`, `nephe_org_list` and
  `nephe_white_list` as
  [defaults](https://jenkins-job-builder.readthedocs.io/en/latest/definition.html#defaults-1s)
  variables in a separate file.

#### Apply the jobs

Run the command to test if jobs can be generated correctly.

```bash
jenkins-jobs test -r ci/jenkins/jobs
```

Run the command to apply these jobs.

```bash
jenkins-jobs update -r ci/jenkins/jobs
```

#### Jenkins job updater

To follow GitOps best practices, there is a job-updater job in Jenkins to detect

> any change in [jobs](./jobs) for every 15 min. As long as a PR to modify code under

Reviewer 2: When a PR containing code changes to jobs[link] dir is merged, Jenkins jobs on cloud will be automatically updated with the new changes.

Author: done

# Nephe CI: Jenkins

## Table of Contents

<!-- toc -->
- [Introduction](#introduction)
- [Jenkins Deployment](#jenkins-deployment)
  - [Teraform](#teraform)
  - [Ansible](#ansible)
  - [Jenkins Job](#jenkins-job)
    - [Requirements](#requirements)
    - [Apply the jobs](#apply-the-jobs)
    - [Jenkins job updater](#jenkins-job-updater)
- [List of jenkins job](#list-of-jenkins-job)
<!-- /toc -->

## Introduction

This directory includes all the scripts required to run CI for nephe.

> We have a static VM configured to run as jenkins node.  When a jenkins job is

Reviewer 2: is a static vm, or a cluster of VMs?

Author: done

triggered, the nephe code will be cloned on the static VM. The jenkins node will
call terraform to a deploy dynamic VM and use ansible for installing any missing

> packages like docker. Then tests will be run on the dynamic VM, the VM will be

Reviewer 2: nit: on that

Author: done

> packages like docker. Then tests will be run on the dynamic VM, the VM will be

Reviewer 2: Rephrase: Upon completion of the test (success or fail), the dynamic VM will be destroyed

Author: done

destroyed once the test succeeds or fails.

## Jenkins Deployment

### Teraform

> We create terrform tfvars file in nephe-ci.sh with jenkins credetials. [deploy.sh](./deploy.sh)

Reviewer 2: Make these bullet points..

> We create terrform tfvars file in nephe-ci.sh with jenkins credetials. [deploy.sh](./deploy.sh)

Reviewer 2: Rephrase: Nephi-ci.sh (link) is responsible for creating terraform tfvars files.

Author: done

is the script to create vm, and store the terraform state in a specified directory,
which can be used to get VM information later. [destroy.sh](./destroy.sh) helps to detroy the VM
and delete terraform workspace as well.

For more details about terraform, please refer [here](https://www.terraform.io/docs).

### Ansible

> nsible will help us for system configuration after the VM installation finishes, all

Reviewer 2: typo

Author: done

> nsible will help us for system configuration after the VM installation finishes, all

Reviewer 2: Rephrase: Ansible is used to configure VM once terraform deploys the dynamic VM. All anisble ..

Author: done

ansible tasks are in playbook/roles. For example, we add task to install docker in
[playbook/roles/docker/tasks/main.yml](playbook/roles/docker/tasks/main.yml).

> Ansile also help to copy ssh keys to dynamic VM, by which we can run tests and collect

Reviewer 2: helps

Reviewer 2: Ansible is used to copy ssh keys to the dynamic VM, which is used to run the test and collect logs from the dynamic VM later.

Author: done

> Ansile also help to copy ssh keys to dynamic VM, by which we can run tests and collect

Reviewer 2: Typo

Author: done

log later. The ssh-keys are generated on static VM if nephe-ci checks there is no
ssh-keys.

For more details about ansible, please refer [here](https://docs.ansible.com/).

### Jenkins Job

> We use jenkins job-builder to manage jenkins jobs, the codes are in the directory jobs.

Reviewer 2: link here

Author: done

#### Requirements

Yaml files under current directory can be generated via
jenkins-job-builder. If you want to try out the tests on your local jenkins

> setup, please notice the following requirements:

Reviewer 2: please check

Author: done

> * Jenkins setup

Reviewer 2: Where do you point to local Jenkins IP/User/Token? Isnt it done via some export.

  * Plugins: ghprb, throttle-concurrents
* Install
  [jenkins-job-builder](https://jenkins-job-builder.readthedocs.io/en/latest/index.html)
* Define your `ANTREA_GIT_CREDENTIAL` which is the credential for your private
  repo
* Define your `ghpr_auth`, `nephe_admin_list`, `nephe_org_list` and
  `nephe_white_list` as
  [defaults](https://jenkins-job-builder.readthedocs.io/en/latest/definition.html#defaults-1s)

>   variables in a separate file

Reviewer 2: nit: .

Author: done

#### Apply the jobs

Run the command to test if jobs can be generated correctly.

```bash
jenkins-jobs test -r ci/jenkins/jobs
```

Run the command to apply these jobs.

```bash
jenkins-jobs update -r ci/jenkins/jobs
```

#### Jenkins job updater

To follow GitOps best practices, there is a job-updater job in Jenkins to detect

> any change in current directory for every 15 min. As long as

Reviewer 2: what directory.. Is it jobs.. explicitly mention the job and link in the repo

Author: done

> any change in current directory for every 15 min. As long as

Author: This is true for antrea, but not for nephe, maybe we can add this later.

Do you think this shall be removed now, and add back when it is available for nephe

a PR to modify code under that path is merged, Jenkins jobs on cloud should be
updated with new code.

> ## List of jenkins job

Reviewer 2: Jenkins Jobs

Author: done

> - `nephe-test-e2e-aws-for-pull-request`: Run end-to-end AWS tests on a Kind cluster with AWS VMs

Reviewer 2: @Anandkumar26 you wanna mention the jobs here? Or you have it added at some other place.

Reviewer: Here is should contain all the jenkins jobs. 
For the contributor, only the trigger phrases are added.
